ws-server/src/server.rs

```diff
@@ -139,49 +143,59 @@ impl<'a> Future for Incoming<'a> {
 	}
 }
 
-async fn handshake(
-	socket: tokio::net::TcpStream,
-	conn_id: ConnectionId,
-	methods: &Methods,
-	cfg: &Settings,
-	stop_monitor: &StopMonitor,
-) -> Result<(), Error> {
+enum Handshake<'a> {
+	Reject { status_code: u16 },
+	Accept { conn_id: ConnectionId, methods: &'a Methods, cfg: &'a Settings, stop_monitor: &'a StopMonitor },
+}
+
+async fn handshake(socket: tokio::net::TcpStream, details: Handshake<'_>) -> Result<(), Error> {
 	// For each incoming background_task we perform a handshake.
 	let mut server = SokettoServer::new(BufReader::new(BufWriter::new(socket.compat())));
 
-	let key = {
-		let req = server.receive_request().await?;
-		let host_check = cfg.allowed_hosts.verify("Host", Some(req.headers().host));
-		let origin_check = cfg.allowed_origins.verify("Origin", req.headers().origin);
-
-		host_check.and(origin_check).map(|()| req.key())
-	};
-
-	match key {
-		Ok(key) => {
-			let accept = Response::Accept { key, protocol: None };
-			server.send_response(&accept).await?;
-		}
-		Err(error) => {
-			let reject = Response::Reject { status_code: 403 };
+	match details {
+		Handshake::Reject { status_code } => {
+			// Forced rejection, don't need to read anything from the socket
+			let reject = Response::Reject { status_code };
 			server.send_response(&reject).await?;
 
-			return Err(error);
+			Ok(())
```

Comment:
what does `Ok(())` vs `Err(error)` imply here? 

I suppose we don't have any error for it in `Incoming` and return `Ok(())` is probably fine because it's already logged once the connection limit is exceeded when `Handshake::Reject` is detected.

Reply:
Yup, returning 429 rejection isn't an error here, it's expected behavior (but we can still get an io error when sending the response for whatever reason).